File: solutions/r/allergies/1/allergies.R
allergy <- function(num) {
  allergies <- c("eggs","peanuts","shellfish","strawberries","tomatoes","chocolate","pollen","cats")
  scores <- rep(FALSE,8)
  base <- 1
  while(num >= 1 && base <= 8) {
    if(num%%2==1) {
      scores[base] <- TRUE
    }
    num <- num %/% 2
    base <- base + 1
  }
  scoring <- as.list(scores)
  names(scoring) <- allergies
  return(scoring)
}

allergic_to <- function(allergy_object, allergy) {
  return(allergy_object[[allergy]])
}

list_allergies <- function(allergy_object) {
  condition <- allergy_object == FALSE
  if(all(condition)) {
    return(character())
  }
  list <- vector("character",length(allergy_object))
  base <- 1
  for(name in names(allergy_object)) {
    if(allergy_object[[name]]) {
      list[base] <- name
      base <- base + 1
    }
  }
  return(list[1:(base-1)])
}

File: solutions/julia/allergies/1/allergies.jl
function allergic_to(score, allergen)
    list = Dict("eggs"=>1,"peanuts"=>2,"shellfish"=>3,"strawberries"=>4,"tomatoes"=>5,"chocolate"=>6,"pollen"=>7,"cats"=>8)
    if haskey(list,allergen)
        last = list[allergen]
    else
        return false
    end
    while last > 1
        score = score ÷ 2
        last -= 1
    end
    return score%2 == 1
end

function allergy_list(score)
    list = ["eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats"]
    output = Set()
    basic = 1
    while score > 0 && basic < 9
        score % 2 == 1 && push!(output,list[basic])
        score = score ÷ 2
        basic += 1
    end
    return output
end
